### ISSUE TITLE
[Not a bug?] PRUNE subquery is not working 

### ISSUE THREAD

**Issue by CurvaXa** [arangodb/arangodb#12372]
Note: if it is not a bug, why it isn't mentioned in the docs and why even AQL validation allows it? It is really confusing and consumes a lot of time to debug.

## My Environment

* __ArangoDB Version__:       3.6.5
* __Storage Engine__:             Default
* __Deployment Mode__:       Single Server
* __Deployment Strategy__:   Ubuntu default
* __Infrastructure__:               Own
* __Operating System__:        Ubuntu 20.04
* __Total RAM in your machine__:        4Gb
* __Disks in use__:        SSD
* __Used Package__:             default Ubuntu

## Component, Query & Data

__Affected feature__: AQL PRUNE
__AQL query (if applicable)__:
```
    FOR u,d IN 1..100 INBOUND 'users/2918324' delegates
    PRUNE  (FOR u1, v1 in INBOUND votingId votes FILTER u1._id == u._id RETURN 1) == [1]
    LET check = (FOR u1, v1 in INBOUND votingId votes FILTER u1._id == u._id RETURN 1) == [1]
    RETURN {u,check}
```

__AQL explain (if applicable)__:
```
Query String (303 chars, cacheable: true):
    LET votingId = 'votings/2918330'
     FOR u,d IN 1..100 INBOUND 'users/2918324' delegates
     PRUNE  (FOR u1, v1 in INBOUND votingId votes FILTER u1._id == u._id RETURN 1) == [1]
     LET check = (FOR u1, v1 in INBOUND votingId votes FILTER u1._id == u._id RETURN 1) == [1]
     RETURN {u,check}

Execution plan:
 Id   NodeType            Est.   Comment
  1   SingletonNode          1   * ROOT
 23   SubqueryStartNode      1     - LET #5 = ( /* subquery begin */
  7   CalculationNode        1       - LET #16 = 1   /* json expression */   /* const assignment */
  4   TraversalNode          1       - FOR u1  /* vertex */ IN 1..1  /* min..maxPathDepth */ INBOUND 'votings/2918330' /* startnode */  votes
  5   CalculationNode        1         - LET #14 = (u1.`_id` == u.`_id`)   /* simple expression */
  6   FilterNode             1         - FILTER #14
 24   SubqueryEndNode        1     - RETURN  #16 ) /* subquery end */
 10   TraversalNode        900     - FOR u  /* vertex */ IN 1..100  /* min..maxPathDepth */ INBOUND 'users/2918324' /* startnode */  delegates
 21   SubqueryStartNode    900       - LET #9 = ( /* subquery begin */
 15   CalculationNode      900         - LET #24 = 1   /* json expression */   /* const assignment */
 12   TraversalNode          1         - FOR u1  /* vertex */ IN 1..1  /* min..maxPathDepth */ INBOUND 'votings/2918330' /* startnode */  votes
 13   CalculationNode        1           - LET #22 = (u1.`_id` == u.`_id`)   /* simple expression */
 14   FilterNode             1           - FILTER #22
 22   SubqueryEndNode        1       - RETURN  #24 ) /* subquery end */
 19   CalculationNode        1       - LET #26 = { "u" : u, "check" : (#9 == [ 1 ]) }   /* simple expression */
 20   ReturnNode             1       - RETURN #26

Indexes used:
 By   Name   Type   Collection   Unique   Sparse   Selectivity   Fields      Ranges
  4   edge   edge   votes        false    false       100.00 %   [ `_to` ]   base INBOUND
 10   edge   edge   delegates    false    false       100.00 %   [ `_to` ]   base INBOUND
 12   edge   edge   votes        false    false       100.00 %   [ `_to` ]   base INBOUND

Traversals on graphs:
 Id  Depth   Vertex collections  Edge collections  Options                                  Filter / Prune Conditions
 4   1..1                        votes             uniqueVertices: none, uniqueEdges: path                           
 10  1..100                      delegates         uniqueVertices: none, uniqueEdges: path  PRUNE (#5 == [ 1 ])      
 12  1..1                        votes             uniqueVertices: none, uniqueEdges: path                           

Optimization rules applied:
 Id   RuleName
  1   move-calculations-up
  2   remove-unnecessary-calculations
  3   optimize-traversals
  4   splice-subqueries
```

__Dataset__:
users
[{"_key":"2918324","_id":"users/2918324","_rev":"_a55olIO---"}
,{"_key":"2918325","_id":"users/2918325","_rev":"_a55olIO--A"},
{"_key":"2918326","_id":"users/2918326","_rev":"_a55olIO--C"},
{"_key":"2918327","_id":"users/2918327","_rev":"_a55olIO--E"}]

delegates
[{"_key":"2918384","_id":"delegates/2918384","_from":"users/2918327","_to":"users/2918324","_rev":"_a55pQyW---"},{"_key":"2918393","_id":"delegates/2918393","_from":"users/2918326","_to":"users/2918327","_rev":"_a55pacO---"},{"_key":"2918397","_id":"delegates/2918397","_from":"users/2918325","_to":"users/2918326","_rev":"_a55pfPa---"}]

votings
[{"_key":"2918330","_id":"votings/2918330","_rev":"_a55olMO---","tag":"a"}]

votes
[{"_key":"2918407","_id":"votes/2918407","_from":"users/2918326","_to":"votings/2918330","_rev":"_a55pruC---"}]

## Steps to reproduce

1. Create data set
2. Run provided AQL

Graph:
![image](https://user-images.githubusercontent.com/3591134/89424307-4a246780-d740-11ea-809a-996d130ed602.png)

__Problem__:
AQL
```
    LET votingId = 'votings/2918330'
     FOR u,d IN 1..100 INBOUND 'users/2918324' delegates
     PRUNE  (FOR u1, v1 in INBOUND votingId votes FILTER u1._id == u._id RETURN 1) == [1]
     LET check = (FOR u1, v1 in INBOUND votingId votes FILTER u1._id == u._id RETURN 1) == [1]
     RETURN {u.Id,check}
```
RETURNS
```
[
  {
    "u": {
      "_key": "2918327",
      "_id": "users/2918327",
      "_rev": "_a55olIO--E"
    },
    "check": false
  },
  {
    "u": {
      "_key": "2918326",
      "_id": "users/2918326",
      "_rev": "_a55olIO--C"
    },
    "check": true
  },
  {
    "u": {
      "_key": "2918325",
      "_id": "users/2918325",
      "_rev": "_a55olIO--A"
    },
    "check": false
  }
]
```
__Expected result__:
```
[
  {
    "u": {
      "_key": "2918327",
      "_id": "users/2918327",
      "_rev": "_a55olIO--E"
    },
    "check": false
  },
  {
    "u": {
      "_key": "2918326",
      "_id": "users/2918326",
      "_rev": "_a55olIO--C"
    },
    "check": true
  }
]
```

Note: please suggest a workaround, if there is no plan to fix it ASAP.
It works like this:
```
LET votingId = 'votings/2918330'
LET allVoted = (FOR u in INBOUND votingId votes  RETURN u._id)
FOR u,d IN 1..100 INBOUND 'users/2918324' delegates
PRUNE  u._id IN allVoted
LET check = (FOR u1, v1 in INBOUND votingId votes FILTER u1._id == u._id RETURN 1) == [1]
RETURN {u,check}
```
But it leads to horrible performance on the big volumes (2-3 mins with 100k users), so it is not an option.
And I can't perform a left join before PRUNE.  So I am really confused.
Best solution that I have for now - create a new collection and insert prefiltered `delegates` into it.

**Comment by goedderz** [comment]
Hi @CurvaXa,

sorry for the late reply. This is a bug in the following sense: Subqueries are not allowed inside a PRUNE condition, because they cannot work there. However, you've run into a case that slipped through. I've opened PRs for this, so this will be fixed soon. Which means in the future, you will get a proper error message that this is not allowed.

This will of course not help you express your query. To discuss other ways this could be done efficiently I probably need to ask you for more information. And this bug report is not the right place for this. Are you perchance already on [our slack](https://slack.arangodb.com/), or willing to join there?

Best,

Tobias

**Comment by Simran-B** [comment]
Looks like this fix will be available in 3.6.7, not 3.6.6.

**Comment by maxkernbach** [comment]
Hi @CurvaXa,

version 3.6.7, which improves the PRUNE error handling, has been released.

From the [Changelog](https://github.com/arangodb/arangodb/blob/edd53d086d812a21858d85152b900af86161e740/CHANGELOG#L96):

> * Fixed some cases where subqueries in PRUNE did not result in a parse error,
>   but either in an incomprehensible error (in 3.7), or undefined behaviour
>   during execution (pre 3.7).

In case you would like to discuss other efficient ways to express your query, please let us know.
